File: plugins/typescript/nginx-lint-plugin/src/config-builder.ts
/**
 * Builds WIT-compatible Config/Directive objects from parser component output.
 *
 * The parser WASM component returns a ParseOutput with an index-based tree
 * representation (to avoid recursive types in WIT). This module reconstructs
 * the method-based Directive/Config interfaces from that flat representation.
 */

import type {
  Config,
  ConfigItem,
  ConfigSnapshot,
  Directive,
  DirectiveContext,
} from "./generated/interfaces/nginx-lint-plugin-config-api.js";
import type {
  DirectiveData,
  ArgumentInfo,
} from "./generated/interfaces/nginx-lint-plugin-data-types.js";
import type { Fix } from "./generated/interfaces/nginx-lint-plugin-types.js";
import { makeIncludeContextMethods } from "./include-context.js";
import type {
  ParseOutput,
  ConfigItem as ParserConfigItem,
  DirectiveContext as ParserDirectiveContext,
} from "../wasm/parser/interfaces/nginx-lint-plugin-parser-types.js";

// ── Wrap DirectiveData with Directive interface ─────────────────────

/** Mirror of the host's `make_range_fix`: a zero-`line` range-based fix. */
function rangeFix(start: number, end: number, newText: string): Fix {
  return {
    line: 0, oldText: undefined, newText,
    deleteLine: false, insertAfter: false,
    startOffset: start, endOffset: end,
  };
}

/** The indent the host prefixes to each inserted line: `" ".repeat(column - 1)`. */
function indentOf(data: DirectiveData): string {
  return " ".repeat(Math.max(data.column - 1, 0));
}

/** Offset of the start of the directive's line, as the host computes it. */
function lineStartOffset(data: DirectiveData): number {
  return Math.max(data.startOffset - (data.column - 1), 0);
}

/**
 * Duck-typed stand-in for the host-backed `config-api.directive` resource.
 *
 * A class rather than an object literal of closures, so that it matches the
 * shape of the real resource: jco generates that as a class, and the literal
 * made `Object.keys()`, spread and destructured methods behave differently
 * depending on which path a plugin's Config came from.
 *
 * Not a performance change, despite the shape suggesting one. Profiling a
 * 10k-line config found per-check time dominated by the component boundary,
 * not by guest-side allocation: lowering the returned LintErrors accounted
 * for over half of it and lifting the snapshot for most of the rest, while
 * everything this class does — walking the tree, allocating the directive
 * objects — came to a few milliseconds in total.
 */
class BuiltDirective implements Directive {
  readonly #data: DirectiveData;
  readonly #resolveBlockItems: () => ConfigItem[];
  readonly #argValues: string[];

  constructor(data: DirectiveData, resolveBlockItems: () => ConfigItem[]) {
    this.#data = data;
    this.#resolveBlockItems = resolveBlockItems;
    this.#argValues = data.args.map((a) => a.value);
  }

  data(): DirectiveData { return this.#data; }
  name(): string { return this.#data.name; }
  is(name: string): boolean { return this.#data.name === name; }
  firstArg(): string | undefined { return this.#argValues[0] ?? undefined; }
  firstArgIs(value: string): boolean { return this.#argValues[0] === value; }
  argAt(index: number): string | undefined { return this.#argValues[index] ?? undefined; }
  lastArg(): string | undefined {
    return this.#argValues.length > 0 ? this.#argValues[this.#argValues.length - 1] : undefined;
  }
  hasArg(value: string): boolean { return this.#argValues.includes(value); }
  argCount(): number { return this.#argValues.length; }
  args(): ArgumentInfo[] { return this.#data.args; }
  line(): number { return this.#data.line; }
  column(): number { return this.#data.column; }
  startOffset(): number { return this.#data.startOffset; }
  endOffset(): number { return this.#data.endOffset; }
  leadingWhitespace(): string { return this.#data.leadingWhitespace; }
  trailingWhitespace(): string { return this.#data.trailingWhitespace; }
  spaceBeforeTerminator(): string { return this.#data.spaceBeforeTerminator; }
  hasBlock(): boolean { return this.#data.hasBlock; }
  blockItems(): ConfigItem[] { return this.#resolveBlockItems(); }
  blockIsRaw(): boolean { return this.#data.blockIsRaw; }

  // The fix constructors below must produce byte-identical Fix records to
  // the host's implementations in src/plugin/component_rule.rs
  // (replace_with / delete_line_fix / insert_* on DirectiveResource), so a
  // plugin gets the same --fix result whether its Config came from the
  // host resource or from a reconstructed snapshot.
  replaceWith(newText: string): Fix {
    const leading = this.#data.leadingWhitespace;
    return rangeFix(
      this.#data.startOffset - leading.length,
      this.#data.endOffset,
      leading + newText,
    );
  }
  deleteLineFix(): Fix {
    return {
      line: this.#data.line, oldText: undefined, newText: "",
      deleteLine: true, insertAfter: false,
      startOffset: undefined, endOffset: undefined,
    };
  }
  insertAfter(newText: string): Fix {
    const offset = this.#data.endOffset;
    return rangeFix(offset, offset, `\n${indentOf(this.#data)}${newText}`);
  }
  insertBefore(newText: string): Fix {
    const offset = lineStartOffset(this.#data);
    return rangeFix(offset, offset, `${indentOf(this.#data)}${newText}\n`);
  }
  insertAfterMany(lines: string[]): Fix {
    const indent = indentOf(this.#data);
    const text = lines.map((line) => `\n${indent}${line}`).join("");
    const offset = this.#data.endOffset;
    return rangeFix(offset, offset, text);
  }
  insertBeforeMany(lines: string[]): Fix {
    const indent = indentOf(this.#data);
    const text = lines.map((line) => `${indent}${line}\n`).join("");
    const offset = lineStartOffset(this.#data);
    return rangeFix(offset, offset, text);
  }
}

function wrapDirective(
  data: DirectiveData,
  resolveBlockItems: () => ConfigItem[],
): Directive {
  return new BuiltDirective(data, resolveBlockItems);
}

// ── Resolve index-based items to ConfigItem tree ────────────────────

function resolveConfigItem(
  allItems: ParserConfigItem[],
  index: number,
): ConfigItem {
  const item = allItems[index];
  if (item.value.tag === "directive-item") {
    const data = item.value.val;
    const childIndices = item.childIndices;
    return {
      tag: "directive-item",
      val: wrapDirective(data, () =>
        Array.from(childIndices).map((i) => resolveConfigItem(allItems, i)),
      ),
    };
  }
  if (item.value.tag === "comment-item") {
    return { tag: "comment-item", val: item.value.val };
  }
  return { tag: "blank-line-item", val: item.value.val };
}

// ── Build Config from ParseOutput ───────────────────────────────────

export function buildConfigFromParseOutput(output: ParseOutput): Config {
  const inclCtx = output.includeContext;
  const allItems = output.allItems;

  const directiveContexts: DirectiveContext[] = output.directivesWithContext.map(
    (ctx: ParserDirectiveContext) => ({
      directive: wrapDirective(ctx.data, () =>
        Array.from(ctx.blockItemIndices).map((i) => resolveConfigItem(allItems, i)),
      ),
      parentStack: ctx.parentStack,
      depth: ctx.depth,
    }),
  );

  const topLevelItems: ConfigItem[] = Array.from(output.topLevelIndices).map(
    (i) => resolveConfigItem(allItems, i),
  );

  return {
    allDirectivesWithContext() { return directiveContexts; },
    allDirectives() { return directiveContexts.map((c) => c.directive); },
    items() { return topLevelItems; },
    snapshot() {
      return { allItems, topLevelIndices: output.topLevelIndices, includeContext: inclCtx };
    },
    snapshotFiltered(names: string[]) {
      const filteredItems: ParserConfigItem[] = [];
      const filteredTopLevel = filterItemsByName(
        allItems,
        Array.from(output.topLevelIndices),
        names,
        filteredItems,
      );
      return {
        allItems: filteredItems,
        topLevelIndices: Uint32Array.from(filteredTopLevel),
        includeContext: inclCtx,
      };
    },
    ...makeIncludeContextMethods(inclCtx),
  } as Config;
}

// ── Filter a flat item tree by directive name (mirrors the host's
// flatten_item_to_wit_filtered in src/plugin/component_rule.rs) ─────
//
// Keeps a directive if its own name is in `names`, OR it has at least one
// kept descendant (so ancestor context like is_inside("http") stays
// correct); comments/blank-lines are always dropped. Used by the parser-
// output-backed test mock's snapshotFiltered() so plugin tests exercise the
// same semantics the real host applies.

function filterItemsByName(
  allItems: ParserConfigItem[],
  indices: number[],
  names: string[],
  outItems: ParserConfigItem[],
): number[] {
  const kept: number[] = [];
  for (const index of indices) {
    const item = allItems[index];
    if (item.value.tag !== "directive-item") continue;
    const data = item.value.val;
    const keptChildIndices = filterItemsByName(
      allItems,
      Array.from(item.childIndices),
      names,
      outItems,
    );
    if (!names.includes(data.name) && keptChildIndices.length === 0) continue;

    const newIndex = outItems.length;
    outItems.push({
      value: { tag: "directive-item", val: data },
      childIndices: Uint32Array.from(keptChildIndices),
    });
    kept.push(newIndex);
  }
  return kept;
}

// ── Build Config from a (possibly filtered) ConfigSnapshot ──────────
//
// Unlike ParseOutput (produced by the parser component for testing),
// ConfigSnapshot (produced by the host's Config.snapshot()/snapshotFiltered())
// does not come with a pre-computed directives-with-context list — the DFS
// walk with parent-stack tracking has to be done here, seeded with
// includeContext exactly as the host does for allDirectivesWithContext().
// Works for both the full snapshot() and a snapshotFiltered(names) result:
// the walk itself doesn't know or care whether the tree was pruned.

function collectDirectiveContexts(
  allItems: ParserConfigItem[],
  indices: number[],
  parentStack: string[],
  depth: number,
  results: DirectiveContext[],
): void {
  for (const index of indices) {
    const item = allItems[index];
    if (item.value.tag !== "directive-item") continue;
    const data = item.value.val;
    const childIndices = Array.from(item.childIndices);

    results.push({
      directive: wrapDirective(data, () =>
        childIndices.map((i) => resolveConfigItem(allItems, i)),
      ),
      parentStack: [...parentStack],
      depth,
    });

    collectDirectiveContexts(
      allItems,
      childIndices,
      [...parentStack, data.name],
      depth + 1,
      results,
    );
  }
}

/**
 * A {@link Config} reconstructed from a snapshot, minus `snapshot()` and
 * `snapshotFiltered()` themselves: {@link buildConfigFromSnapshot} doesn't
 * implement them (there's no live host resource behind a reconstructed
 * config to re-fetch from), so calling either on the result would throw at
 * runtime. Omitting them from the type makes that a compile-time error
 * instead of a runtime surprise for a plugin that tries to re-filter an
 * already-filtered config.
 */
export type ReconstructedConfig = Omit<Config, "snapshot" | "snapshotFiltered">;

export function buildConfigFromSnapshot(snapshot: ConfigSnapshot): ReconstructedConfig {
  const inclCtx = snapshot.includeContext;
  const allItems = snapshot.allItems;
  const topLevelIndices = Array.from(snapshot.topLevelIndices);

  const directiveContexts: DirectiveContext[] = [];
  collectDirectiveContexts(
    allItems,
    topLevelIndices,
    inclCtx,
    inclCtx.length,
    directiveContexts,
  );

  const topLevelItems: ConfigItem[] = topLevelIndices.map((i) =>
    resolveConfigItem(allItems, i),
  );

  return {
    allDirectivesWithContext() { return directiveContexts; },
    allDirectives() { return directiveContexts.map((c) => c.directive); },
    items() { return topLevelItems; },
    ...makeIncludeContextMethods(inclCtx),
  } as ReconstructedConfig;
}

File: plugins/python/nginx-lint-plugin/nginx_lint_plugin/config_builder.py
"""Builds WIT-compatible Config/Directive objects from parser output.

Port of the TypeScript SDK's config-builder.ts. The parser returns a
ParseOutput with an index-based tree representation (to avoid recursive
types in WIT); this module reconstructs the method-based Directive/Config
interfaces from that flat representation, with the exact method names the
componentize-py generated bindings use (``is_``, ``first_arg_is``, ...),
so a plugin's ``check()`` runs unmodified against the result.
"""

from typing import Callable, List, Optional

from wit_world.imports import config_api, parser_types
from wit_world.imports.config_api import ConfigSnapshot
from wit_world.imports.data_types import ArgumentInfo, DirectiveData
from wit_world.imports.parser_types import ParseOutput
from wit_world.imports.types import Fix

from .include_context import IncludeContextMixin

# ── Directive wrapper ───────────────────────────────────────────────


class BuiltDirective:
    """Duck-typed stand-in for the host-backed ``config-api.directive`` resource."""

    def __init__(
        self,
        data: DirectiveData,
        resolve_block_items: Callable[[], List[config_api.ConfigItem]],
    ):
        self._data = data
        self._resolve_block_items = resolve_block_items
        self._arg_values = [a.value for a in data.args]

    def data(self) -> DirectiveData:
        return self._data

    def name(self) -> str:
        return self._data.name

    def is_(self, name: str) -> bool:
        return self._data.name == name

    def first_arg(self) -> Optional[str]:
        return self._arg_values[0] if self._arg_values else None

    def first_arg_is(self, value: str) -> bool:
        return bool(self._arg_values) and self._arg_values[0] == value

    def arg_at(self, index: int) -> Optional[str]:
        return self._arg_values[index] if 0 <= index < len(self._arg_values) else None

    def last_arg(self) -> Optional[str]:
        return self._arg_values[-1] if self._arg_values else None

    def has_arg(self, value: str) -> bool:
        return value in self._arg_values

    def arg_count(self) -> int:
        return len(self._arg_values)

    def args(self) -> List[ArgumentInfo]:
        return self._data.args

    def line(self) -> int:
        return self._data.line

    def column(self) -> int:
        return self._data.column

    def start_offset(self) -> int:
        return self._data.start_offset

    def end_offset(self) -> int:
        return self._data.end_offset

    def leading_whitespace(self) -> str:
        return self._data.leading_whitespace

    def trailing_whitespace(self) -> str:
        return self._data.trailing_whitespace

    def space_before_terminator(self) -> str:
        return self._data.space_before_terminator

    def has_block(self) -> bool:
        return self._data.has_block

    def block_items(self) -> List[config_api.ConfigItem]:
        return self._resolve_block_items()

    def block_is_raw(self) -> bool:
        return self._data.block_is_raw

    def replace_with(self, new_text: str) -> Fix:
        return Fix(
            line=self._data.line,
            old_text=None,
            new_text=new_text,
            delete_line=False,
            insert_after=False,
            start_offset=self._data.start_offset,
            end_offset=self._data.end_offset,
        )

    def delete_line_fix(self) -> Fix:
        return Fix(
            line=self._data.line,
            old_text=None,
            new_text="",
            delete_line=True,
            insert_after=False,
            start_offset=None,
            end_offset=None,
        )

    def insert_after(self, new_text: str) -> Fix:
        return Fix(
            line=self._data.line,
            old_text=None,
            new_text=new_text,
            delete_line=False,
            insert_after=True,
            start_offset=None,
            end_offset=None,
        )

    def insert_before(self, new_text: str) -> Fix:
        return Fix(
            line=self._data.line,
            old_text=None,
            new_text=new_text,
            delete_line=False,
            insert_after=False,
            start_offset=None,
            end_offset=None,
        )

    def insert_after_many(self, lines: List[str]) -> Fix:
        return Fix(
            line=self._data.line,
            old_text=None,
            new_text="\n".join(lines),
            delete_line=False,
            insert_after=True,
            start_offset=None,
            end_offset=None,
        )

    def insert_before_many(self, lines: List[str]) -> Fix:
        return Fix(
            line=self._data.line,
            old_text=None,
            new_text="\n".join(lines),
            delete_line=False,
            insert_after=False,
            start_offset=None,
            end_offset=None,
        )


# ── Resolve index-based items to ConfigItem tree ────────────────────


def _resolve_config_item(
    all_items: List[parser_types.ConfigItem], index: int
) -> config_api.ConfigItem:
    item = all_items[index]
    value = item.value
    if isinstance(value, parser_types.ConfigItemValue_DirectiveItem):
        child_indices = list(item.child_indices)
        return config_api.ConfigItem_DirectiveItem(
            value=BuiltDirective(
                value.value,
                lambda: [_resolve_config_item(all_items, i) for i in child_indices],
            )
        )
    if isinstance(value, parser_types.ConfigItemValue_CommentItem):
        return config_api.ConfigItem_CommentItem(value=value.value)
    return config_api.ConfigItem_BlankLineItem(value=value.value)


# ── Config implementations ──────────────────────────────────────────


class _BaseConfig(IncludeContextMixin):
    def __init__(
        self,
        directive_contexts: List[config_api.DirectiveContext],
        top_level_items: List[config_api.ConfigItem],
        include_context: List[str],
    ):
        self._directive_contexts = directive_contexts
        self._top_level_items = top_level_items
        self._include_context = include_context

    def all_directives_with_context(self) -> List[config_api.DirectiveContext]:
        return self._directive_contexts

    def all_directives(self) -> List[BuiltDirective]:
        return [c.directive for c in self._directive_contexts]

    def items(self) -> List[config_api.ConfigItem]:
        return self._top_level_items


class BuiltConfig(_BaseConfig):
    """Duck-typed stand-in for the host-backed ``config-api.config`` resource."""

    def __init__(
        self,
        directive_contexts: List[config_api.DirectiveContext],
        top_level_items: List[config_api.ConfigItem],
        output: ParseOutput,
    ):
        super().__init__(directive_contexts, top_level_items, output.include_context)
        self._output = output

    def snapshot(self) -> ConfigSnapshot:
        return ConfigSnapshot(
            all_items=self._output.all_items,
            top_level_indices=self._output.top_level_indices,
            include_context=self._output.include_context,
        )

    def snapshot_filtered(self, names: List[str]) -> ConfigSnapshot:
        filtered_items: List[parser_types.ConfigItem] = []
        filtered_top_level = _filter_items_by_name(
            self._output.all_items,
            list(self._output.top_level_indices),
            names,
            filtered_items,
        )
        return ConfigSnapshot(
            all_items=filtered_items,
            top_level_indices=filtered_top_level,
            include_context=self._output.include_context,
        )


class ReconstructedConfig(_BaseConfig):
    """A Config reconstructed from a snapshot.

    Unlike :class:`BuiltConfig` it has no ``snapshot()``/``snapshot_filtered()``:
    there's no live host resource behind a reconstructed config to re-fetch
    from, so a plugin that tries to re-filter an already-filtered config gets
    an AttributeError instead of silently wrong data.
    """


# ── Build Config from ParseOutput ───────────────────────────────────


def build_config_from_parse_output(output: ParseOutput) -> BuiltConfig:
    all_items = output.all_items

    directive_contexts = [
        config_api.DirectiveContext(
            directive=BuiltDirective(
                ctx.data,
                (lambda indices: lambda: [
                    _resolve_config_item(all_items, i) for i in indices
                ])(list(ctx.block_item_indices)),
            ),
            parent_stack=list(ctx.parent_stack),
            depth=ctx.depth,
        )
        for ctx in output.directives_with_context
    ]

    top_level_items = [
        _resolve_config_item(all_items, i) for i in output.top_level_indices
    ]

    return BuiltConfig(directive_contexts, top_level_items, output)


# ── Filter a flat item tree by directive name (mirrors the host's
# flatten_item_to_wit_filtered in src/plugin/component_rule.rs) ─────
#
# Keeps a directive if its own name is in `names`, OR it has at least one
# kept descendant (so ancestor context like is_inside("http") stays
# correct); comments/blank-lines are always dropped.


def _filter_items_by_name(
    all_items: List[parser_types.ConfigItem],
    indices: List[int],
    names: List[str],
    out_items: List[parser_types.ConfigItem],
) -> List[int]:
    kept: List[int] = []
    for index in indices:
        item = all_items[index]
        if not isinstance(item.value, parser_types.ConfigItemValue_DirectiveItem):
            continue
        data = item.value.value
        kept_child_indices = _filter_items_by_name(
            all_items, list(item.child_indices), names, out_items
        )
        if data.name not in names and not kept_child_indices:
            continue

        new_index = len(out_items)
        out_items.append(
            parser_types.ConfigItem(
                value=parser_types.ConfigItemValue_DirectiveItem(value=data),
                child_indices=kept_child_indices,
            )
        )
        kept.append(new_index)
    return kept


# ── Build Config from a (possibly filtered) ConfigSnapshot ──────────
#
# ConfigSnapshot does not come with a pre-computed directives-with-context
# list — the DFS walk with parent-stack tracking is done here, seeded with
# include_context exactly as the host does for all_directives_with_context().


def _collect_directive_contexts(
    all_items: List[parser_types.ConfigItem],
    indices: List[int],
    parent_stack: List[str],
    depth: int,
    results: List[config_api.DirectiveContext],
) -> None:
    for index in indices:
        item = all_items[index]
        if not isinstance(item.value, parser_types.ConfigItemValue_DirectiveItem):
            continue
        data = item.value.value
        child_indices = list(item.child_indices)

        results.append(
            config_api.DirectiveContext(
                directive=BuiltDirective(
                    data,
                    (lambda ci: lambda: [
                        _resolve_config_item(all_items, i) for i in ci
                    ])(child_indices),
                ),
                parent_stack=list(parent_stack),
                depth=depth,
            )
        )

        _collect_directive_contexts(
            all_items,
            child_indices,
            parent_stack + [data.name],
            depth + 1,
            results,
        )


def build_config_from_snapshot(snapshot: ConfigSnapshot) -> ReconstructedConfig:
    incl_ctx = list(snapshot.include_context)
    all_items = snapshot.all_items
    top_level_indices = list(snapshot.top_level_indices)

    directive_contexts: List[config_api.DirectiveContext] = []
    _collect_directive_contexts(
        all_items,
        top_level_indices,
        incl_ctx,
        len(incl_ctx),
        directive_contexts,
    )

    top_level_items = [_resolve_config_item(all_items, i) for i in top_level_indices]

    return ReconstructedConfig(directive_contexts, top_level_items, incl_ctx)
